File: services/ai-service/src/services/telegramService.js
const Redis = require('ioredis');

const redisClient = new Redis({
  host: process.env.REDIS_HOST || 'kiora-redis',
  port: Number(process.env.REDIS_PORT) || 6379,
});

const REDIS_STREAM = process.env.REDIS_NOTIFICATIONS_STREAM || 'kiora:notifications:stream';
const HISTORY_EXPIRATION = 86400; // 24 horas

async function saveTelegramHistory(chatId, role, content) {
  const key = `kiora:telegram_history:${chatId}`;
  await redisClient.rpush(key, JSON.stringify({ role, content }));
  // Mantener solo los últimos 20 mensajes
  await redisClient.ltrim(key, -20, -1);
  await redisClient.expire(key, HISTORY_EXPIRATION);
}

async function getTelegramHistory(chatId) {
  const key = `kiora:telegram_history:${chatId}`;
  const messages = await redisClient.lrange(key, 0, -1);
  return messages.map(msg => JSON.parse(msg));
}

async function sendTelegramNotification(subject, htmlContent) {
  try {
    const payload = JSON.stringify({ subject, html: htmlContent });
    await redisClient.xadd(REDIS_STREAM, '*', 'payload', payload);
    return true;
  } catch (error) {
    console.error('[telegramService] Error enviando notificación:', error.message);
    return false;
  }
}

module.exports = {
  saveTelegramHistory,
  getTelegramHistory,
  sendTelegramNotification,
  redisClient
};

File: services/ai-service/app/services/telegram_service.py
import os
import json
import logging
import redis.asyncio as redis
from datetime import datetime

logger = logging.getLogger(__name__)

REDIS_HOST = os.getenv("REDIS_HOST", "kiora-redis")
REDIS_PORT = int(os.getenv("REDIS_PORT", "6379"))
REDIS_STREAM = os.getenv("REDIS_NOTIFICATIONS_STREAM", "kiora:notifications:stream")
HISTORY_EXPIRATION = 86400  # 24 hours

redis_client = redis.Redis(host=REDIS_HOST, port=REDIS_PORT, decode_responses=True)

async def save_telegram_history(chat_id: str, role: str, content: str):
    try:
        key = f"kiora:telegram_history:{chat_id}"
        msg = json.dumps({"role": role, "content": content})
        await redis_client.rpush(key, msg)
        # Keep only the last 20 messages
        await redis_client.ltrim(key, -20, -1)
        await redis_client.expire(key, HISTORY_EXPIRATION)
    except Exception as e:
        logger.error(f"[TelegramService] Error saving history: {e}")

async def get_telegram_history(chat_id: str) -> list:
    try:
        key = f"kiora:telegram_history:{chat_id}"
        messages_raw = await redis_client.lrange(key, 0, -1)
        messages = []
        for msg in messages_raw:
            try:
                messages.append(json.loads(msg))
            except:
                pass
        return messages
    except Exception as e:
        logger.error(f"[TelegramService] Error getting history: {e}")
        return []

async def send_telegram_notification(subject: str, html: str, photo_base64: str = None):
    try:
        payload = {
            "subject": subject,
            "html": html,
            "timestamp": datetime.utcnow().isoformat() + "Z"
        }
        if photo_base64:
            payload["photo_base64"] = photo_base64
            
        await redis_client.xadd(
            "kiora:notifications:stream",
            {"payload": json.dumps(payload)}
        )
        return True
    except Exception as e:
        logger.error(f"[Telegram Notification] Error sending to Redis stream: {e}")
        return False
